components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/openidconnect/JWTAccessTokenOIDCClaimsHandler.java

```diff
@@ -1016,4 +1026,202 @@ private boolean isPreserverClaimUrisInAssertion(OAuthTokenReqMessageContext requ
+                                               String spTenantDomain, String clientId) {
+        try {
+            // Check if roles claim is present in the claims (using OIDC dialect)
+            if (claims.containsKey(ROLES)) {
+                AuthenticatedUser authenticatedUser = requestMsgCtx.getAuthorizedUser();
+                ServiceProvider serviceProvider = getServiceProvider(spTenantDomain, clientId);
+                
+                if (serviceProvider != null && authenticatedUser != null) {
+                    String applicationId = serviceProvider.getApplicationResourceId();
+                    
+                    // Get application-associated roles for the user
+                    String[] appAssociatedRoles = OIDCClaimUtil.getAppAssociatedRolesOfUser(
+                            authenticatedUser, applicationId);
+                    
+                    if (appAssociatedRoles != null && appAssociatedRoles.length > 0) {
+                        // Filter out null values and join with separator
+                        String[] validRoles = Arrays.stream(appAssociatedRoles)
+                                .filter(Objects::nonNull)
+                                .toArray(String[]::new);
+                        
+                        if (validRoles.length > 0) {
+                            String rolesValue = String.join(FrameworkUtils.getMultiAttributeSeparator(), validRoles);
+                            claims.put(ROLES, rolesValue);
+                            
+                            if (log.isDebugEnabled()) {
+                                log.debug("Filtered roles for application " + clientId + ": " + 
+                                        String.join(", ", validRoles));
+                            }
+                        } else {
+                            // No valid roles, set empty value
+                            claims.put(ROLES, "");
+                            if (log.isDebugEnabled()) {
+                                log.debug("No valid application-specific roles found for user " + 
+                                        authenticatedUser.getLoggableMaskedUserId() + " in application " + clientId);
+                            }
+                        }
+                    } else {
+                        // No app-associated roles found, set empty value to prevent all user roles
+                        claims.put(ROLES, "");
+                        if (log.isDebugEnabled()) {
+                            log.debug("No application-associated roles found for user " + 
+                                    authenticatedUser.getLoggableMaskedUserId() + " in application " + clientId);
+                        }
+                    }
+                }
+            }
+            
+            // Also handle APP_ROLES_CLAIM if present
+            if (claims.containsKey(APP_ROLES)) {
```

Comment:
I think there is no possibility for this scenario since we only add app_roles claim with filtered list